### ISSUE TITLE
Build Ceres(1.12) for Android with openmp, but seems not speed up

### ISSUE THREAD

**Issue by PotterSu** [ceres-solver/ceres-solver#498]
Hello bodys,
I just have built ceres(1.12.0) with openmp、suitesparse by Android.mk.
And I solve my problem by trust_region_strategy_type.
Here's my options:
m_options.trust_region_strategy_type = ceres::LEVENBERG_MARQUARDT;
m_options.max_trust_region_radius = 1e26;
m_options.initial_trust_region_radius = 1e26;
m_options.minimizer_progress_to_stdout = true;
m_options.num_threads = 2;

Now,I test my program on Android phone. But when I set num_threads to 1 or 2 or 4, the time is almost the same. I wonder if openmp is supported by Android platform.
Here's my full report:
Threds = 1 Report:
Solver Summary (v 1.12.0-eigen-(3.3.7)-lapack-suitesparse-(5.3.0)-openmp)

                                     Original                  Reduced
Parameter blocks                         8032                     8032
Parameters                              16064                    16064
Residual blocks                         32513                    32513
Residual                                65026                    65026

Minimizer                        TRUST_REGION

Sparse linear algebra library    SUITE_SPARSE

Trust region strategy     LEVENBERG_MARQUARDT

                                        Given                     Used
Linear solver          SPARSE_NORMAL_CHOLESKY   SPARSE_NORMAL_CHOLESKY
Threads                                     1                        1
Linear solver threads                       1                        1
Linear solver ordering              AUTOMATIC                     8032

Cost:
Initial                          1.427145e+07
Final                            4.792502e+06
Change                           9.478948e+06

Minimizer iterations                        4
Successful steps                            4
Unsuccessful steps                          0

Time (in seconds):
Preprocessor                           0.0481

  Residual evaluation                  0.0204
  Jacobian evaluation                  0.1157
  Linear solver                        0.3030
Minimizer                              0.4699

Postprocessor                          0.0016
Total                                  0.5196

Termination:                      CONVERGENCE (Function tolerance reached. |cost_change|/cost: 2.190923e-07 <= 1.000000e-06)

Threads = 2 Report:
Solver Summary (v 1.12.0-eigen-(3.3.7)-lapack-suitesparse-(5.3.0)-openmp)

                                     Original                  Reduced
Parameter blocks                         8032                     8032
Parameters                              16064                    16064
Residual blocks                         32513                    32513
Residual                                65026                    65026

Minimizer                        TRUST_REGION

Sparse linear algebra library    SUITE_SPARSE

Trust region strategy     LEVENBERG_MARQUARDT

                                        Given                     Used
Linear solver          SPARSE_NORMAL_CHOLESKY   SPARSE_NORMAL_CHOLESKY
Threads                                     2                        2
Linear solver threads                       1                        1
Linear solver ordering              AUTOMATIC                     8032

Cost:
Initial                          1.427145e+07
Final                            4.792502e+06
Change                           9.478948e+06

Minimizer iterations                        4
Successful steps                            4
Unsuccessful steps                          0

Time (in seconds):
Preprocessor                           0.0458

  Residual evaluation                  0.0212
  Jacobian evaluation                  0.1169
  Linear solver                        0.3071
Minimizer                              0.4782

Postprocessor                          0.0018
Postprocessor                          0.0018
Total                                  0.5257

Termination:                      CONVERGENCE (Function tolerance reached. |cost_change|/cost: 2.190923e-07 <= 1.000000e-06)

There's another question:
I changed and built Android Ceres with CERES_USE_CXX11_THREADS. I built success but when I opened 4 threads, it also could not speed up my program. The full report says 
 Given                     Used
Linear solver          SPARSE_NORMAL_CHOLESKY   SPARSE_NORMAL_CHOLESKY
Threads                                     4                        1

So, is ceres-1.12.0 support multi-threadings speed up or I should built master branches ceres?
Best wishes and Hope for your responds!

**Comment by sandwichmaker** [comment]
1.12 is very very old. Could you try the master branch and report back please?

**Comment by PotterSu** [comment]
@sandwichmaker 
Thank you for your reply!
**So, is v1.12 is supported for multi-threads or not? Or 1.12 is supported for multi-threads but speed up is not very obvious? As I showed, threads given 2 and used 2 indeed but time used is almost the same.**

Ok, I compiled master branch v2.0.0 without suite-sparse.
I test ceres's example-- robust_curve_fitting.cc using DENSE_QR.
**When num_threads = 1, total time is about 0.002815s, sometimes around 0.001.**
Solver Summary (v 2.0.0-eigen-(3.3.7)-no_lapack-eigensparse-no_openmp)

                                     Original                  Reduced
Parameter blocks                            2                        2
Parameters                                  2                        2
Residual blocks                            67                       67
Residuals                                  67                       67

Minimizer                        TRUST_REGION


Dense linear algebra library            EIGEN
Trust region strategy     LEVENBERG_MARQUARDT

                                        Given                     Used
Linear solver                        DENSE_QR                 DENSE_QR
**Threads                                     1                        1**
Linear solver ordering              AUTOMATIC                        2

Cost:
Initial                          1.815138e+01
Final                            1.902884e+00
Change                           1.624849e+01

Minimizer iterations                       12
Successful steps                            8
Unsuccessful steps                          4
Time (in seconds):
Preprocessor                         0.000257

  Residual only evaluation           0.000169 (12)
  Jacobian & residual evaluation     0.000234 (8)
  Linear solver                      0.000078 (12)
Minimizer                            0.002549

Postprocessor                        0.000009
**Total                                0.002815**

Termination:                      CONVERGENCE (Function tolerance reached. |cost_change|/cost: 5.631229e-07 <= 1.000000e-06)

Initial m: 0 c: 0
Final   m: 0.287605 c: 0.151213

**When num_threads = 2, total time is about 0.0051s, I test several times and total times are all increased .**
                                        Given                     Used
Linear solver                        DENSE_QR                 DENSE_QR
**Threads                                     2                        2**
Linear solver ordering              AUTOMATIC                        2

Cost:
Initial                          1.815138e+01
Final                            1.902884e+00
Change                           1.624849e+01

Minimizer iterations                       12
Successful steps                            8
Unsuccessful steps                          4
Time (in seconds):
Preprocessor                         0.000340

  Residual only evaluation           0.001931 (12)
  Jacobian & residual evaluation     0.001055 (8)
  Linear solver                      0.000092 (12)
Minimizer                            0.004811

Postprocessor                        0.000009
**Total                                0.005163**

Termination:                      CONVERGENCE (Function tolerance reached. |cost_change|/cost: 5.631229e-07 <= 1.000000e-06)

Initial m: 0 c: 0
Final   m: 0.287605 c: 0.151213

**When threads = 4, total time is up to 0.024s**
 Given                     Used
Linear solver                        DENSE_QR                 DENSE_QR
Threads                                     4                        4
Linear solver ordering              AUTOMATIC                        2

Cost:
Initial                          1.815138e+01
Final                            1.902884e+00
Change                           1.624849e+01


Minimizer iterations                       12
Successful steps                            8
Unsuccessful steps                          4

Time (in seconds):
Preprocessor                         0.006018

  Residual only evaluation           0.002782 (12)
  Jacobian & residual evaluation     0.013012 (8)
  Linear solver                      0.000205 (12)
Minimizer                            0.018390

Postprocessor                        0.000019
Total                                0.024428

Termination:                      CONVERGENCE (Function tolerance reached. |cost_change|/cost: 5.631229e-07 <= 1.000000e-06)

Initial m: 0 c: 0
Final   m: 0.287605 c: 0.151213

My android phone's cpu is Qualcomm Snapdragon 710.
When I test the same example(robust_curve_fitting.cc) on older version 1.12.0. Whatever I set threads to 1 or 2 or 4, cost time is almost the same about 0.024s. 
So, is this normal? Open multi-threads will cause more time and this program is too simple to open multi-threads? 
Today I'll try v2.0.0 compiled with suite-sparse and test my program again.
 Thank you!

**Comment by sandwichmaker** [comment]
The curve fitting problem is very small. the big change between 1.12 and 1.14 was that instead of using openmp thread directly we started using an abstraction which can use openmp or c++ threads. But getting multithreading to work well on android is hard due to heterogeneous cores and how the processor does thread scheduling. 

Either ways the curve fitting problem is rather small to be indicative of threading performance. When you are solving something with SuiteSparse, the only thing that will get parallelized is the evaluation of Jacobians and residuals. This should show in your Summary:FullReport.

**Comment by PotterSu** [comment]
OK,  I'll try thank you! 
Maybe compiling v2.0.0(Android platform) with CMakeLists is a little difficult for me. For the reason that suite-sparse downloaded from website is not supported for Android very well. They are organized by makefiles not CMakeLists...

**Comment by PotterSu** [comment]
@sandwichmaker 
Hi,I'm back again~
I compiled ceres v2.0.0(master branch) with suite-sparse, and I found when I set num_threads to 2 costs much more time than num_threads=1.
**And I also found when I set CERES_NO_THREADS, and I also can set num_threads to 2. The cost time is also more than 1 thread. There seems that -DCERES_NO_THREADS or -DCERES_USE_CXX11_THREADS not used.**

**Here's threads = 1 Report:**
iter      cost      cost_change  |gradient|   |step|    tr_ratio  tr_radius  ls_iter  iter_time  total_time
   0  5.932409e+06    0.00e+00    1.47e+06   0.00e+00   0.00e+00  1.00e+26        0    2.16e-02    6.75e-02

CHOLMOD version 3.0.9, Feb 1, 2016: Symbolic Analysis: status: OK
  Architecture: Linux
    sizeof(int):      4
    sizeof(SuiteSparse_long):  4
    sizeof(void *):   4

    sizeof(double):   8
    sizeof(Int):      4 (CHOLMOD's basic integer)
    sizeof(BLAS_INT): 4 (integer used in the BLAS)
  Results from most recent analysis:
    Cholesky flop count: 5.9855e+07
    Nonzeros in L:       6.025e+05
  memory blocks in use:          11
  memory in use (MB):           0.9
  peak memory usage (MB):       1.5
  maxrank:    update/downdate rank:   8
  supernodal control: 1 40 (supernodal if flops/lnz >= 40)
  nmethods:   number of ordering methods to try: 1
    method 0: natural
        flop count: 5.9855e+07
        nnz(L):     6.025e+05
  OK
   1  4.129927e+06    1.80e+06    2.27e+02   5.43e+03   9.94e-01  1.00e+26        1    1.20e-01    1.88e-01
   2  4.119435e+06    1.05e+04    4.51e+00   1.92e+02   1.00e+00  1.00e+26        1    9.17e-02    2.81e-01
MTWideAngleUnDistortion: debug ObjectTermSolver[971]: Solve problem cost time = 368.948000!

Solver Summary (v 2.0.0-eigen-(3.3.7)-no_lapack-suitesparse-(5.3.0)-eigensparse-no_openmp)

                                     Original                  Reduced
Parameter blocks                         8032                     8032
Parameters                              16064                    16064
Residual blocks                         32529                    32529
Residuals                               65058                    65058


Minimizer                        TRUST_REGION

Sparse linear algebra library    SUITE_SPARSE
Trust region strategy     LEVENBERG_MARQUARDT

                                        Given                     Used
Linear solver          SPARSE_NORMAL_CHOLESKY   SPARSE_NORMAL_CHOLESKY
Threads                                     1                        1
Linear solver ordering              AUTOMATIC                     8032

Cost:
Initial                          5.932409e+06
Final                            4.119435e+06
Change                           1.812974e+06

Minimizer iterations                        3
Successful steps                            3
Unsuccessful steps                          0

**Time (in seconds):**
Preprocessor                         0.045917
Residual only evaluation           0.014554 (3)
  Jacobian & residual evaluation     0.049000 (3)
  Linear solver                      0.222654 (3)
Minimizer                            0.311629

Postprocessor                        0.001855
**Total                                0.359404**

Termination:                      CONVERGENCE (Function tolerance reached. |cost_change|/cost: 6.637136e-07 <= 1.000000e-06)

**Here's threads = 2 Report**
iter      cost      cost_change  |gradient|   |step|    tr_ratio  tr_radius  ls_iter  iter_time  total_time
   0  5.932409e+06    0.00e+00    1.47e+06   0.00e+00   0.00e+00  1.00e+26        0    5.78e-02    1.03e-01

CHOLMOD version 3.0.9, Feb 1, 2016: Symbolic Analysis: status: OK
  Architecture: Linux
    sizeof(int):      4
    sizeof(SuiteSparse_long):  4
    sizeof(void *):   4

    sizeof(double):   8
    sizeof(Int):      4 (CHOLMOD's basic integer)
    sizeof(BLAS_INT): 4 (integer used in the BLAS)
  Results from most recent analysis:
    Cholesky flop count: 5.9855e+07
    Nonzeros in L:       6.025e+05
  memory blocks in use:          11
  memory in use (MB):           0.9
  peak memory usage (MB):       1.5
  maxrank:    update/downdate rank:   8
  supernodal control: 1 40 (supernodal if flops/lnz >= 40)
  nmethods:   number of ordering methods to try: 1
    method 0: natural
        flop count: 5.9855e+07
        nnz(L):     6.025e+05
  OK
   1  4.129927e+06    1.80e+06    2.27e+02   5.43e+03   9.94e-01  1.00e+26        1    2.16e-01    3.19e-01
   2  4.119435e+06    1.05e+04    4.51e+00   1.92e+02   1.00e+00  1.00e+26        1    2.15e-01    5.37e-01
MTWideAngleUnDistortion: debug ObjectTermSolver[971]: Solve problem cost time = 762.188000!

Solver Summary (v 2.0.0-eigen-(3.3.7)-no_lapack-suitesparse-(5.3.0)-eigensparse-no_openmp)
Original                  Reduced
Parameter blocks                         8032                     8032
Parameters                              16064                    16064
Residual blocks                         32529                    32529
Residuals                               65058                    65058

Minimizer                        TRUST_REGION

Sparse linear algebra library    SUITE_SPARSE

Trust region strategy     LEVENBERG_MARQUARDT

                                        Given                     Used
Linear solver          SPARSE_NORMAL_CHOLESKY   SPARSE_NORMAL_CHOLESKY
Threads                                     2                        2
Linear solver ordering              AUTOMATIC                     8032

Cost:
Initial                          5.932409e+06
Final                            4.119435e+06
Change                           1.812974e+06

Minimizer iterations                        3
Successful steps                            3
Unsuccessful steps                          0

**Time (in seconds):**
Preprocessor                         0.044812

  Residual only evaluation           0.135976 (3)
  Jacobian & residual evaluation     0.179287 (3)
  Linear solver                      0.305697 (3)
Minimizer                            0.689054
Postprocessor                        0.006700
**Total                                0.740570**

Termination:                      CONVERGENCE (Function tolerance reached. |cost_change|/cost: 6.637136e-07 <= 1.000000e-06)

I think the problem to be solved is complex enough and it's strange that more threads and more cost time.
Thanks for your reply!
